### DISCUSSION TITLE
switching from Dirichlet to Neumann BC when a certain condition is met

### DISCUSSION THREAD

**Discussion by nlosacco** [idaholab/moose#26203] [Q&A General]
Hi all,

I'm trying to simulate runoff on a slope in a simplified way as asked here: https://github.com/idaholab/moose/discussions/25591 , but I want to pose a more general question about switching between D and N BCs.
The physical concept is:
1) an initially partly saturated slope (porewater pressure u < 0 at the top);
2) increasing rainfall, simulated aș a flow from the top, causes u to increase until u = 0, then the condition u = 0 must be applied to simulate the water film at the top;
3) as the imposed flow reduces, simulating periods of less intense rainfall or evapotranspiration (i.e. negative flow), so that u<0 would be predicted, the condition u = 0 at the top must be removed, switching back to a prescribed flow at the top.

Before simulating this quite complex phenomenon in detail I am trying to understand how to tackle it with MOOSE, by simulating steady-state diffusion flow in a simple, saturated column, in order to validate my approach.
I have prescribed D-BC u = 0 at the bottom and applied time-varying N-BC at the top (through a `ParsedFunction`) so that, theoretically according to Darcy's law, u at the top should vary from 0 to 20 and back to 0 during the analysis.
Now, I want to limit u<=10 at the top so my plan is:
1) use a `ConditionalFunctionEnable` to check wether the u = 10 at the top
2) if, during the calculation, u > 10 is predicted due the imposed N-BC at the top, then switch to a `PenaltyDirichletBC` to impose u=10 at the top;
3) at the beginning of the following calculation step, the time-varying N-BC should be applied again at the top and the check described at step 2) should be carried out.

The question is:
I understand I have to perform the check at step 2) through a `ConditionalFunctionEnableControl` , what I am wandering, though, is what is the best point in the calculation scheme to perform this check. The real problem which I will tackle when I solve this, involving partly saturated flow and mechanical coupling, is strongly non linear. I think I should perform the check on the predicted u at each iteration (i.e. before calculating residual), with `execute_on = LINEAR`. But what happens when the linear predictor at a certain iteration yields u>10? Is the whole time-step aborted and re-started with the new BC? Is that single iteration aborted and then repeated according to the penalty scheme? I think the latter should be the way to go, because theoretically speaking the condition might change during the iterations of a given time-step, but how does MOOSE deal with it?
Finally, I am willing to use a penalty approach because I'm afraid switching from N-BC to D-BC might impose excessive non-linearity to the problem, making it harder to solve.

Sorry for the TLRD post and thanks for you patience and guidance.

**Comment by josebastiase** [comment]
Hi, 
interesting problem. Can you provide a minimal example of your soil column so we have starting point?

**Comment by nlosacco** [comment]
This is the simplest example I can think of, no PorousFlow, just basic diffusion. @vabufano
```
[Mesh]
     type = GeneratedMesh
     dim = 2
     nx = 2
     ny = 10
     xmax = 2.
     ymax = 10.
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
[]

[Functions]
  [top_pres]
     type=ParsedFunction
     expression = 2.0*t
  []
[]

[BCs]
  [top]
    type = FunctionNeumannBC
    variable = u
    boundary = top
    function = top_pres
#   value = 1.
  []
  [bottom]
    type = DirichletBC
    variable = u
    boundary = bottom
    value = 0.
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'hypre'
  dt = 0.1

[Postprocessors]
  [top_pp]
    type = PointValue
    point = '1. 10. 0.'
    variable = u
  []
[]

[Outputs]
  [out]
     type = Exodus
  []
[]
```

**Comment by GiudGiud** [comment]
> I think I should perform the check on the predicted u at each iteration (i.e. before calculating residual), with execute_on = LINEAR. But what happens when the linear predictor at a certain iteration yields u>10? Is the whole time-step aborted and re-started with the new BC?

The whole time step is not restarted. The BC is changed for the rest of the linear iterations. On the next nonlinear iteration, you start another set of linear iterations so the criteria is checked again. It could bounce between BCs, which would be the problem with executing it on LINEAR. TIMESTEP_BEGIN would not bounce within a timestep, but it could bounce between time steps. 

> Is that single iteration aborted and then repeated according to the penalty scheme? I think the latter should be the way to go, but how does MOOSE deal with it?

Nope the penalty scheme can be turned on anytime. MOOSE reforms a new residual and a new Jacobian on the respective iterations, and we can change what contributes to these objects at any time. 
But actually, you're making me think that for a Newton scheme, if you change the BC on linear, the Jacobian would not be updated until the next nonlinear iteration.

**Reply by nlosacco** [reply]
@GiudGiud thank for your enlightening answer. No problem to me if it bounces between BCs during a single time-step, as far as it eventually converges. According to what you wrote, I'll try executing on both NONLINEAR and LINEAR and check how the results and convergence behaviour change in the complete case (i.e. with unsaturated soil, PorousFlow, etc.)

**Reply by nlosacco** [reply]
Hi, I tried with the following code:

```
[Mesh]
     type = GeneratedMesh
     dim = 2
     nx = 2
     ny = 10
     xmax = 2.
     ymax = 10.
[]

[Variables]
  [u]
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
[]

[Functions]
  [top_pres]
     type=ParsedFunction
     expression = 2.0*t
  []
  [top_pres_condition]
     type=ParsedFunction
     expression = 'u_sol >= 15.'
     symbol_names = 'u_sol'
     symbol_values = 'top_pp'
  []
[]

[BCs]
  [top]
    type = FunctionNeumannBC
    variable = u
    boundary = top
    function = top_pres
#   value = 1.
  []
  [top_penalty]
     type = PenaltyDirichletBC
     variable = u
     boundary = top
     value = 15.
     penalty = 1e3
  []
  [bottom]
    type = DirichletBC
    variable = u
    boundary = bottom
    value = 0.
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'hypre'
  dt = 0.01
  end_time = 1.0
[]

[Controls]
  [max_pwp]
     type = ConditionalFunctionEnableControl
     conditional_function = 'top_pres_condition'
     enable_objects = 'BCs::top_penalty'
     disable_objects = 'BCs::top'
     execute_on = 'INITIAL TIMESTEP_BEGIN LINEAR NONLINEAR TIMESTEP_END'
  []
[]

[Postprocessors]
  [top_pp]
    type = PointValue
    point = '1. 10. 0.'
    variable = u
  []
[]

[Outputs]
  [out]
     type = Exodus
  []
[]
```

and here's the results for the last bit of the analysis:

```
+----------------+----------------+
| time           | top_pp         |
+----------------+----------------+
:                :                :
|   3.000000e-01 |   6.000000e+00 |
|   3.500000e-01 |   7.000000e+00 |
|   4.000000e-01 |   8.000000e+00 |
|   4.500000e-01 |   9.000000e+00 |
|   5.000000e-01 |   1.000000e+01 |
|   5.500000e-01 |   1.100000e+01 |
|   6.000000e-01 |   1.200000e+01 |
|   6.500000e-01 |   1.300000e+01 |
|   7.000000e-01 |   1.400000e+01 |
|   7.500000e-01 |   1.500000e+01 |
|   8.000000e-01 |   1.600000e+01 |
|   8.500000e-01 |   1.499850e+01 |
|   9.000000e-01 |   1.800000e+01 |
|   9.500000e-01 |   1.499850e+01 |
|   1.000000e+00 |   2.000000e+01 |
+----------------+----------------+
```

As a matter of fact, from t = 0.8 onwards, the solution should yield (approximately) 15. constantly.
I assume the reason is that I am using the `Postprocessor` value in the conditional `ParsedFunction`, which I assume is only available at the end of a time-step. Checks, instead, should be done with respect to a value that is changing during the iterations.
I am a bit confused, what can I try next?

**Reply by GiudGiud** [reply]
It looks like it's bouncing right?
So top_pp > 15 causes the change, which causes top_pp < 15 so the change is undone.

If you set `reverse_on_false = false` to the control, it will not be undone. The action of the control will only happen once to flip the BCs.
Is that what you want?

**Reply by nlosacco** [reply]
Er… what I want to achieve is limiting the value of the solution variable to a maximum. Theoretically, the condition should be checked at all times during the iterations. So, no, `reverse_on_false = false` would not do.
My gut feeling is that `Postprocessors` are only evaluated at the end of the time step, in fact the penalty BC seems to work well: it prescribes a value slightly smaller than the threshold, this is why at the following timestamp, and every other one after that, based on the previously calculated post processor the penalty is turned off.
I feel an AuxVariable would do, as I understand they are evaluated during the iterations, just like ordinary variables. But I didn’t try yet. Do you think this might work?

**Reply by GiudGiud** [reply]
you can change the execution schedule of Postprocessors using the execute_on parameter

**Comment by nlosacco** [comment]
I did this, but I get the same solution :-(
How can I print the postprocessor value for each and every iteration?

```
[Controls]
  [max_pwp]
     type = ConditionalFunctionEnableControl
     conditional_function = 'top_pres_condition'
     enable_objects = 'BCs::top_penalty'
     disable_objects = 'BCs::top'
     execute_on = 'INITIAL TIMESTEP_BEGIN LINEAR NONLINEAR TIMESTEP_END'
  []
[]

[Postprocessors]
  [top_pp]
    type = PointValue
    point = '1. 10. 0.'
    variable = u
    execute_on = 'INITIAL TIMESTEP_BEGIN LINEAR NONLINEAR TIMESTEP_END'
  []
[]
```

**Reply by GiudGiud** [reply]
the CSV output can be performed on linear and nonlinear. Add it to your Outputs block

**Reply by nlosacco** [reply]
[test_csv.csv](https://github.com/idaholab/moose/files/13540086/test_csv.csv)
Hi all,

not entirely sure how, but playing with penalty factor, absolute tolerance and controls, I kinda made it. Please see the pasted code. Now, if I use Dt = 0.05, u = 15. is achieved exactly at t = 7.5, and everything works hunky dory. If I use Dt = 0.1, then u = 16 at t = 8, and only after that the procedure is working correctly:

I also attach the the csv I produced in output. At time 0.7002 (I understood it is the beginning of the second non-linear iteration) top_pp = 16, but aux_variable flag (which yields the result of the conditional function) does not become 1 until the second linear iteration. In any case, even if the condition is true, the penalty BC is not activated for that timestep. It is activated correctly from the following onwards. It is like the Control is only performed at the beginning of the timestep, regrdless of the corresponding "execute_on" parameter.
What am I missing?

```
Postprocessor Values:
+----------------+----------------+
| time           | top_pp         |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
|   1.000000e-01 |   2.000000e+00 |
|   2.000000e-01 |   4.000000e+00 |
|   3.000000e-01 |   6.000000e+00 |
|   4.000000e-01 |   8.000000e+00 |
|   5.000000e-01 |   1.000000e+01 |
|   6.000000e-01 |   1.200000e+01 |
|   7.000000e-01 |   1.400000e+01 |
|   8.000000e-01 |   1.600000e+01 |
|   9.000000e-01 |   1.499985e+01 |
|   1.000000e+00 |   1.499985e+01 |
+----------------+----------------+
``` 

```
[Mesh]
     type = GeneratedMesh
     dim = 2
     nx = 2
     ny = 10
     xmax = 2.
     ymax = 10.
[]

[Variables]
  [u]
  []
[]

[AuxVariables]
  [flag]
     family = SCALAR
  []
[]

[AuxScalarKernels]
  [flag_aux]
     type = FunctionScalarAux
     variable = flag
     function = top_pres_condition
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
  []
[]

[Functions]
  [top_pres]
     type=ParsedFunction
     expression = 2.0*t
  []
  [top_pres_condition]
     type=ParsedFunction
     expression = 'u_sol > 14.999'
     symbol_names = 'u_sol'
     symbol_values = 'top_pp'
  []
[]

[BCs]
  [top]
    type = FunctionNeumannBC
    variable = u
    boundary = top
    function = top_pres 
#   value = 1.
  []

  [top_penalty]
     type = PenaltyDirichletBC
     variable = u
     boundary = top
     value = 15.000
     penalty = 1e4
  []

  [bottom]
    type = DirichletBC
    variable = u
    boundary = bottom
    value = 0.
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'hypre'
  dt = 0.1
  end_time = 1.0
  nl_abs_tol = 1.E-8
#  automatic_scaling = true
[]

[Controls]
  [max_pwp]
     type = ConditionalFunctionEnableControl
     conditional_function = 'top_pres_condition'
     enable_objects = 'BCs::top_penalty'
     disable_objects = 'BCs::top'
    execute_on = 'TIMESTEP_BEGIN NONLINEAR LINEAR TIMESTEP_END'
  []
[]

[Postprocessors]
  [top_pp]
    type = PointValue
    point = '1. 10. 0.'
    variable = u
    execute_on = 'TIMESTEP_BEGIN NONLINEAR LINEAR TIMESTEP_END'
  []
[]

[Outputs]
  [cons]
    type = CSV
    execute_on = 'TIMESTEP_BEGIN LINEAR NONLINEAR TIMESTEP_END'
  []
  [out]
     type = Exodus
  []
[] 
```

**Reply by nlosacco** [reply]
I ain't sure, but it really seems that some controls are only activated either at TIMESTEP_BEGIN or TIMESTEP_END. I tried to print it to the console:

```
  [contolout]
    type = ControlOutput
    show_active_objects = False
    execute_on = 'TIMESTEP_BEGIN LINEAR NONLINEAR TIMESTEP_END'
  []
```
I don't  get the Control status printed, but at the beginning and end of the timestep:

```
Time Step 8, time = 0.8, dt = 0.1

Active Controls:
    BCs/top/enable = 1 <bool>
    BCs/top_penalty/enable = 0 <bool>

 0 Nonlinear |R| = 2.449490e-01
      0 Linear |R| = 2.449490e-01
      1 Linear |R| = 1.083005e-02
      2 Linear |R| = 2.968707e-04
      3 Linear |R| = 1.844288e-06
 1 Nonlinear |R| = 1.975459e-06
      0 Linear |R| = 1.975459e-06
      1 Linear |R| = 3.575992e-08
      2 Linear |R| = 8.643607e-10
      3 Linear |R| = 1.342318e-11
 2 Nonlinear |R| = 1.350652e-11
 Solve Converged!

Active Controls:
    BCs/top/enable = 0 <bool>
    BCs/top_penalty/enable = 1 <bool>
```

please check also the csv attached to my. previous message

**Reply by GiudGiud** [reply]
Well the print of the status may only be on timestep begin and end, it does not mean controls are not executed on linear and nonlinear. You could add a log in the control to check. Otherwise the FEProblemBase & NLSystemBase classes can be inspected for the calls to execute

**Reply by nlosacco** [reply]
thanks @GiudGiud ,I tried and checked FEProblemBase and NonlinearSytemBase but, due to my complete ignorance of C++ I don't have a clue about what to do. I am quite stuck at this point but BC is not switched during the timestep when it should, i.e. when any iteration triggers the condition. Furthermore, the DirichletBC should be re-activated at the beginning of each timestep.

**Reply by nlosacco** [reply]
Can you tell me how to add a log to Control?
Also, I thought maybe controls are executed but BCs are not updated during the timestep. Can you point me to the functions in the source code that pick the BCs for Jacobian and residual calculation? I might give a try.

**Reply by GiudGiud** [reply]
In the code for Control object, add a print statement. Then you can see when it gets executed. 

The code that selects object is within the residual loops (in src/loops/ComputeResidualThread and its base class mostly)
then in the calls to onSubdomainChanged, we select objects (here kernels) that are active on each block. There's something similar for BCs. This is when objects are checked for their 'active' status.

**Reply by lindsayad** [reply]
Do you allow u>10 in any other part of your domain? If not, I think a variational inequality formulation would be better for your setup than this attempt to switch objects in the middle of a Newton iteration. You would implement this with the `Bounds` system, and then some PETSc options
```
petsc_options_iname = '-snes_type'
petsc_options_value = 'vinewtonrsls'
```
or instead of `vinewtwonrsls` you can use `vinewtonssls`. You could look at [this test input](https://github.com/idaholab/moose/blob/next/test/tests/nodalkernels/constraint_enforcement/vi-bounding.i) as an example

**Reply by lindsayad** [reply]
We only update active objects at the beginning of timesteps. That seems to be the problem you are encountering. The relevant logic is in `FixedPointSolve::solveStep`

**Reply by nlosacco** [reply]
Hi @lindsayad 
`Bounds`sounds great. Alas, in the real problem I only need to limit u at the top boundary of the mesh and, as far as I understand, the `Bounds` system applies to the whole domain. Hence, it seems I might only try and tweak the code, although I have no idea... Apart my specific needs, of limiting rainfall infiltration in a slope, I think this functionality would be important for many use cases

**Reply by lindsayad** [reply]
Well at this point I would say your best best is to write your own integrated boundary condition which switches (using a good old C++ `if` statement) between the penalty Dirichlet and Neumann conditions as you require. Discontinuities like these in the residual function are tough on Newton, but if that's your physics, then that's what we've got to deal with

**Reply by nlosacco** [reply]
> Discontinuities like these in the residual function are tough on Newton, but if that's your physics, then that's what we've got to deal with

I agree but consider that during the application of the Neumann BC (flux at the boundary) the limit value should be approached slowly and the subsequent penalty approach should help.
I don't know if something like `Bounds` can be manipulated so as to be applied only at the boundary. That would be great

**Reply by lindsayad** [reply]
> I don't know if something like Bounds can be manipulated so as to be applied only at the boundary

Not at this time. How we would typically handle this kind of thing is to implement the variational inequality directly in the residual object. 

But I do think the easiest thing here is in your own integrated BC just switch between Neumann and penalty Dirichlet as those are the conditions you want to model. The penalty condition is an alternative implementation to what a semi-smooth variational inequality solver might do with a Lagrange multiplier to enforce the variable bound

**Comment by richmondodufisan** [comment]
I thought this was a pretty interesting problem and I wanted to suggest an alternative method. What if you use the terminator:

https://mooseframework.inl.gov/source/userobjects/Terminator.html

to terminate the solve when the condition is met, and then use the restart and recover feature

https://mooseframework.inl.gov/application_usage/restart_recover.html

to continue from the last stopping point, this time with Neumann conditions? It might be slightly more work to implement and you'd probably need two input files + a bash/python script to automate switching between the two, but I think it might be conceptually easier to do.

**Reply by nlosacco** [reply]
Hi @richmondodufisan ,
thanks a lot for the suggestion, I'll give a try. But consider that this is something that might occur hundreds of times during a simulation, when, for instance, you are simulating the rainfall history on a hillslope over tens of years. I think the approach you suggested is a bit unpractical in this scenario, but it might be very useful for simpler cases. I'll let you know!
